File: java/src/main/java/io/binac/leetcode/WordLadderII.java
package io.binac.leetcode;

import java.util.*;

/**
 * <p>A <strong>transformation sequence</strong> from word <code>beginWord</code> to word <code>endWord</code> using a dictionary <code>wordList</code> is a sequence of words <code>beginWord -&gt; s<sub>1</sub> -&gt; s<sub>2</sub> -&gt; ... -&gt; s<sub>k</sub></code> such that:</p>
 *
 * <ul>
 * 	<li>Every adjacent pair of words differs by a single letter.</li>
 * 	<li>Every <code>s<sub>i</sub></code> for <code>1 &lt;= i &lt;= k</code> is in <code>wordList</code>. Note that <code>beginWord</code> does not need to be in <code>wordList</code>.</li>
 * 	<li><code>s<sub>k</sub> == endWord</code></li>
 * </ul>
 *
 * <p>Given two words, <code>beginWord</code> and <code>endWord</code>, and a dictionary <code>wordList</code>, return <em>all the <strong>shortest transformation sequences</strong> from</em> <code>beginWord</code> <em>to</em> <code>endWord</code><em>, or an empty list if no such sequence exists. Each sequence should be returned as a list of the words </em><code>[beginWord, s<sub>1</sub>, s<sub>2</sub>, ..., s<sub>k</sub>]</code>.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 *
 * <pre><strong>Input:</strong> beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]
 * <strong>Output:</strong> [["hit","hot","dot","dog","cog"],["hit","hot","lot","log","cog"]]
 * <strong>Explanation:</strong>&nbsp;There are 2 shortest transformation sequences:
 * "hit" -&gt; "hot" -&gt; "dot" -&gt; "dog" -&gt; "cog"
 * "hit" -&gt; "hot" -&gt; "lot" -&gt; "log" -&gt; "cog"
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log"]
 * <strong>Output:</strong> []
 * <strong>Explanation:</strong> The endWord "cog" is not in wordList, therefore there is no valid transformation sequence.
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>1 &lt;= beginWord.length &lt;= 5</code></li>
 * 	<li><code>endWord.length == beginWord.length</code></li>
 * 	<li><code>1 &lt;= wordList.length &lt;= 1000</code></li>
 * 	<li><code>wordList[i].length == beginWord.length</code></li>
 * 	<li><code>beginWord</code>, <code>endWord</code>, and <code>wordList[i]</code> consist of lowercase English letters.</li>
 * 	<li><code>beginWord != endWord</code></li>
 * 	<li>All the words in <code>wordList</code> are <strong>unique</strong>.</li>
 * </ul>
 */
public class WordLadderII {
    public static class Solution1 {
        private void addToGraph(Map<String, Set<String>> graph, String parent, String child, boolean reverse) {
            if (reverse) {
                String t = parent;
                parent = child;
                child = t;
            }
            Set<String> children = graph.computeIfAbsent(parent, k -> new HashSet<>());
            children.add(child);
        }

        private void findLadders(Map<String, Set<String>> tree, String begin, String end,
                                 List<List<String>> result, List<String> path) {
            if (begin.equals(end)) {
                result.add(new ArrayList<>(path));
                return;
            }
            Set<String> children = tree.get(begin);
            if (children == null)
                return;
            for (String child : children) {
                path.add(child);
                findLadders(tree, child, end, result, path);
                path.remove(path.size() - 1);
            }
        }

        public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
            List<List<String>> result = new ArrayList<>();

            Set<String> dict = new HashSet<>(wordList);
            dict.remove(beginWord);
            if (!dict.remove(endWord)) return result;
            dict.remove(beginWord);

            Set<String> set1 = new HashSet<>();
            set1.add(beginWord);
            Set<String> set2 = new HashSet<>();
            set2.add(endWord);
            Set<String> next = new HashSet<>();

            Map<String, Set<String>> graph = new HashMap<>();
            boolean reverse = false;

            final int width = beginWord.length();

            boolean found = false;
            while (!found && !set1.isEmpty() && !set2.isEmpty()) {
                if (set1.size() > set2.size()) {
                    Set<String> s = set1;
                    set1 = set2;
                    set2 = s;
                    reverse = !reverse;
                }

                for (String str : set1) {
                    char[] chars = str.toCharArray();
                    for (int j = 0; j < width; ++j) {
                        char origin = chars[j];
                        for (char ch = 'a'; ch <= 'z'; ++ch) {
                            if (ch == origin) continue;
                            chars[j] = ch;
                            String s = String.valueOf(chars);
                            if (set2.contains(s)) {
                                found = true;
                                addToGraph(graph, str, s, reverse);
                            } else if (!found && dict.contains(s)) {
                                next.add(s);
                                addToGraph(graph, str, s, reverse);
                            }
                        }
                        chars[j] = origin;
                    }
                }
                dict.removeAll(next);
                {
                    Set<String> s = set1;
                    set1 = next;
                    next = s;
                    next.clear();
                }
            }

            if (!found)
                return result;
            List<String> path = new ArrayList<>();
            path.add(beginWord);
            findLadders(graph, beginWord, endWord, result, path);
            return result;
        }
    }
}

File: python3/leetcodepy/word_ladder_ii.py
"""
A transformation sequence from word beginWord to word endWord using a dictionary wordList is a sequence of words beginWord -> s1 -> s2 -> ... -> sk such that:

Every adjacent pair of words differs by a single letter.
Every si for 1 <= i <= k is in wordList. Note that beginWord does not need to be in wordList.
sk == endWord

Given two words, beginWord and endWord, and a dictionary wordList, return all the shortest transformation sequences from beginWord to endWord, or an empty list if no such sequence exists. Each sequence should be returned as a list of the words [beginWord, s1, s2, ..., sk].



Example 1:

Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]
Output: [["hit","hot","dot","dog","cog"],["hit","hot","lot","log","cog"]]
Explanation: There are 2 shortest transformation sequences:
"hit" -> "hot" -> "dot" -> "dog" -> "cog"
"hit" -> "hot" -> "lot" -> "log" -> "cog"

Example 2:

Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log"]
Output: []
Explanation: The endWord "cog" is not in wordList, therefore there is no valid transformation sequence.


Constraints:

1 <= beginWord.length <= 5
endWord.length == beginWord.length
1 <= wordList.length <= 1000
wordList[i].length == beginWord.length
beginWord, endWord, and wordList[i] consist of lowercase English letters.
beginWord != endWord
All the words in wordList are unique.
"""
import string
from collections import defaultdict
from typing import DefaultDict, List, Set


class Solution1:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        result = []
        word_set = set(wordList)
        if endWord not in word_set:
            return result
        word_set.remove(endWord)
        word_set.discard(beginWord)

        width = len(beginWord)
        s1, s2 = {beginWord}, {endWord}
        found = False
        reverse = False
        graph = defaultdict(set)
        while not found and s1 and s2:
            if len(s1) > len(s2):
                s1, s2 = s2, s1
                reverse = not reverse
            _next = set()
            for word in s1:
                for i in range(width):
                    left, right = word[:i], word[i + 1:]
                    old = word[i]
                    for ch in string.ascii_lowercase:
                        if ch == old:
                            continue
                        guess = left + ch + right
                        if guess in s2:
                            found = True
                            self._graph_add(graph, word, guess, reverse)
                        elif not found and guess in word_set:
                            _next.add(guess)
                            self._graph_add(graph, word, guess, reverse)
            word_set -= _next
            s1 = _next
        if not found:
            return result
        self._graph_dfs(graph, beginWord, endWord, result, [beginWord])
        return result

    def _graph_add(self, graph: DefaultDict[str, Set[str]], parent: str, child: str, flip: bool):
        if flip:
            graph[child].add(parent)
        else:
            graph[parent].add(child)

    def _graph_dfs(self, graph: DefaultDict[str, Set[str]], word: str, end_word: str,
                   result: List[List[str]], path: List[str]):
        if word == end_word:
            result.append(path[:])
            return
        for s in graph[word]:
            path.append(s)
            self._graph_dfs(graph, s, end_word, result, path)
            path.pop()
